sinks/opentsdb/driver.go

```diff
@@ -0,0 +1,234 @@
+	return opentsdbSinkName
+}
+
+// timeSeriesToPoint transfers the contents holding in the given pointer of sink_api.Timeseries
+// into the instance of opentsdbclient.DataPoint
+func (tsdbSink *openTSDBSink) timeSeriesToPoint(timeseries *sink_api.Timeseries) opentsdbclient.DataPoint {
+	seriesName := strings.Replace(timeseries.Point.Name, "/", "_", -1)
+	if timeseries.MetricDescriptor.Units.String() != "" {
+		seriesName = fmt.Sprintf("%s_%s", seriesName, timeseries.MetricDescriptor.Units.String())
+	}
+	if timeseries.MetricDescriptor.Type.String() != "" {
+		seriesName = fmt.Sprintf("%s_%s", seriesName, timeseries.MetricDescriptor.Type.String())
+	}
+	datapoint := opentsdbclient.DataPoint{
+		Metric:    seriesName,
+		Tags:      make(map[string]string, len(timeseries.Point.Labels)),
+		Value:     timeseries.Point.Value,
+		Timestamp: timeseries.Point.End.Unix(),
+	}
+	for key, value := range timeseries.Point.Labels {
+		if value != "" {
+			datapoint.Tags[key] = value
+		}
+	}
+
+	tsdbSink.secureTags(&datapoint)
+	return datapoint
+}
+
+// eventToPoint transfers the contents holding in the given pointer of sink_api.Event
+// into the instance of opentsdbclient.DataPoint
+func (tsdbSink *openTSDBSink) eventToPoint(event *kube_api.Event) *opentsdbclient.DataPoint {
+	datapoint := opentsdbclient.DataPoint{
+		Metric: eventMetricName,
+		Tags: map[string]string{
+			eventUID:  string(event.UID),
+			eventHost: event.Source.Host,
+		},
+		Timestamp: event.LastTimestamp.Time.Unix(),
+	}
+	if valueStr, err := getEventValue(event); err == nil {
+		datapoint.Value = valueStr
+	} else {
+		glog.Errorf("error occurs when trying to get event info - %v", err)
+		return nil
+	}
+
+	if event.InvolvedObject.Kind == "Pod" {
+		datapoint.Tags[eventPodID] = string(event.InvolvedObject.UID)
+		datapoint.Tags[eventPodName] = event.InvolvedObject.Name
+	}
+	return &datapoint
+}
+
+// Generate point value for event
+func getEventValue(event *kube_api.Event) (string, error) {
+	bytes, err := json.MarshalIndent(event, "", " ")
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+// secureTags just fills in the default key-value pair for the tags, if there is no tags for
+// current datapoint. Otherwise, the opentsdb will return error and the operation of putting
+// datapoint will be failed.
+func (tsdbSink *openTSDBSink) secureTags(datapoint *opentsdbclient.DataPoint) {
+	if len(datapoint.Tags) == 0 {
+		datapoint.Tags[defaultTagName] = defaultTagValue
+	}
+}
+
+func (tsdbSink *openTSDBSink) recordWriteFailure() {
+	tsdbSink.Lock()
+	defer tsdbSink.Unlock()
+	tsdbSink.writeFailures++
+}
+
+func (tsdbSink *openTSDBSink) getState() string {
+	tsdbSink.RLock()
+	defer tsdbSink.RUnlock()
+	return fmt.Sprintf("\tNumber of write failures: %d\n", tsdbSink.writeFailures)
+}
+
+func new(opentsdbHost string) (sink_api.ExternalSink, error) {
+	cfg := opentsdbcfg.OpenTSDBConfig{OpentsdbHost: opentsdbHost}
+	opentsdbClient, err := opentsdbclient.NewClient(cfg)
```

Comment:
'opentsdbclient.NewClient(cfg)' currently returns error only in the scene that there is no value of cfg.OpentsdbHost. 

So in this quoted stage, there is no error if currently the tsdb server isn't available.
Each time calling StoreTimeseries() or StoreEvents(), it will use tsdbSink.client.Ping() to check
whether the opentsdb server is available. This will make sure push datapoints to opentsdb when it is available in most cases.
```java
func (tsdbSink *openTSDBSink) StoreTimeseries(timeseries []sink_api.Timeseries) error {
	if timeseries == nil || len(timeseries) <= 0 {
		return nil
	}
	if err := tsdbSink.client.Ping(); err != nil {
		return err
	}
	dataPoints := make([]opentsdbclient.DataPoint, 0, len(timeseries))
	for _, series := range timeseries {
		dataPoints = append(dataPoints, tsdbSink.timeSeriesToPoint(&series))
	}
	_, err := tsdbSink.client.Put(dataPoints, opentsdbclient.PutRespWithSummary)
	if err != nil {
		glog.Errorf("failed to write timeseries to opentsdb - %v", err)
		tsdbSink.recordWriteFailure()
		return err
	}
	return nil
}

func (tsdbSink *openTSDBSink) StoreEvents(events []kube_api.Event) error {
	if events == nil || len(events) <= 0 {
		return nil
	}
	if err := tsdbSink.client.Ping(); err != nil {
		return err
	}
	dataPoints := make([]opentsdbclient.DataPoint, 0, len(events))
	for _, event := range events {
		datapointPtr := tsdbSink.eventToPoint(&event)
		if datapointPtr != nil {
			dataPoints = append(dataPoints, *datapointPtr)
		}
	}
	_, err := tsdbSink.client.Put(dataPoints, opentsdbclient.PutRespWithSummary)
	if err != nil {
		glog.Errorf("failed to write events to opentsdb - %v", err)
		tsdbSink.recordWriteFailure()
		return err
	}
	return nil
}
```